docs/architecture/adr-006-02-client-refactor.md

```diff
@@ -0,0 +1,170 @@
+As @seunlanlege [states](https://github.com/cosmos/ibc-go/issues/284#issuecomment-1005583679):
+```
+I'm in support of these changes for 2 reasons:
+
+-  This would allow light clients handle batch header updates in CheckHeaderAndUpdateState, for the special case of 11-beefy proving the finality for a batch of headers is much more space and time efficient than the space/time complexity of proving each individual headers in that batch, combined.
+
+-  This also allows for a single light client instance of 11-beefy be used to prove finality for every parachain connected to the relay chain (Polkadot/Kusama). We achieve this by setting the appropriate ConsensusState for individual parachain headers in CheckHeaderAndUpdateState
+```
+
+## Decision
+
+### Require light clients to set client and consensus states
+The IBC specification states:
+```
+If the provided header was valid, the client MUST also mutate internal state to store now-finalised consensus roots and update any necessary signature authority tracking (e.g. changes to the validator set) for future calls to the validity predicate.
+```
+
+The initial version of the IBC go SDK based module did not fulfill this requirement. 
+Instead, the 02-client submodule required each light client to return the client and consensus state which should be updated in the client prefixed store. 
+This decision lead to the issues "Solomachine doesn't set consensus states" and "New clients may want to do batch updates". 
+
+Each light client should be required to set its own client and consensus states on any update necessary. 
+The go implementation should be changed to match the specification requirements. 
+This will allow more flexibility for light clients to manage their own internal storage and do batch updates. 
+
+### Split CheckHeaderAndUpdateState into 4 functions
+
+See [#668](https://github.com/cosmos/ibc-go/issues/668).
+
+Split CheckHeaderAndUpdateState into 4 functions:
+- VerifyClientMessage
+- CheckForMisbehaviour
+- UpdateStateOnMisbehaviour
+- UpdateState
+
+VerifyClientMessage: checks the that the structure of a ClientMessage is correct and that all authentication data provided is valid.
+
+CheckForMisbehaviour: checks to see if a ClientMessage is evidence of misbehaviour.
+
+UpdateStateOnMisbehaviour: freeze the client and updates its state accordingly
+
+UpdateState: Perform a regular update or a no-op on duplicate updates
+
+The code roughly looks like:
+```go
+    if err := clientState.VerifyClientMessage(clientMessage); err != nil {
+        return err
+    }
+    
+    foundMisbehaviour := clientState.CheckForMisbehaviour(clientMessage)
+    if foundMisbehaviour {
+        clientState.UpdateStateOnMisbehaviour(header)
+        // emit misbehaviour event
+        return 
+    }
+    
+    clientState.UpdateState(clientMessage) // expects no-op on duplicate header
+    // emit update event
+    return
+}
+```
+
+### Merge Header/Misbehaviour interface and rename to ClientMessage
```

Comment:
This is the first place where you explain what `ClientMessage` is, right? But you have mentioned it already before when talking about the functions functions splitted from `CheckHeaderAndUpdateState`. Would it make sense to have this section first then?

Reply:
updated